In [1]:
import math
import numpy as np
from scipy.stats import norm
import scipy.stats as st
import matplotlib.pyplot as plt 
import seaborn as sb
import statistics

# Q1
Weiss' paint store stocks a particularly popular white latex paint, whose monthly demand is approximately normal with mean 28 and standard deviation of 8. Replenishment lead time is about 14 weeks. Each can of paint costs the store \\$6. Although excess demands are back- ordered, the store owner estimates that unfilled demands cost about $10 each in bookkeeping and loss-of-goodwill costs. Fixed costs of replenishment are \\$15 per order, and holding costs are based on a 30% annual rate of interest. Assume that there are 4.33 weeks per month.

(a) Suppose that the white paint is reordered on a monthly basis based on the inventory counts at the end of each month. Using the (Q,R) solution, determine the appropriate values of the (s, S) inventory control policy.

In [35]:
# Given 

# Mean = 28
# standard deviation = 8

tau = 14  # Replenishment lead time in weeks which is approx 3.5 months
c = 6     # cost of each paint can
p = 10    # stock out cost per unit unsatisfied demand (book keeping and loss-of-goodwill costs)
K = 15    # setup cost per order
h = 0.3*6 # Holding costs

# Lead time in a year = l
# Decision variables, (Q,R),  Q = order quantity and R = reorder point
# Assumption = 4.33 weeks per month

d = 12 * 28   # mean rate of demand
u = d*tau/52
var_annual = 12*8*8
var_ltd = var_annual * tau/52

In [36]:
# Iteration 0

Q0 = np.sqrt(2*K*d/h)
FR_0 = 1 - ((Q0*h)/(p*d))
z0 = st.norm.ppf(FR_0)         # From appendix normal table
R0 = np.sqrt(var_ltd) * z0 + u
Lz0 = 0.0162
nR0 = Lz0 * np.sqrt(var_ltd)

print('The optimal quantity for iteration 1 is',Q0)
print('The value of F(R_0) is',FR_0)
print('The z-value would be',z0)
print('The R_0 value would be',R0)
print('The value of normal loss function would be',nR0)

The optimal quantity for iteration 1 is 74.83314773547883
The value of F(R_0) is 0.9599108137131364
The z-value would be 1.749652048970506
The R_0 value would be 115.62061206890337
The value of normal loss function would be 0.2329474552835401


In [37]:
# Iteration 2

Q1 = np.sqrt((2*d*K + 2*d*p*nR0)/h)
FR_1 = 1 - ((Q1*h)/(p*d))
z1 = st.norm.ppf(FR_1)         # From normal table
R1 = np.sqrt(var_ltd) * z1 + u

print('The optimal quantity for iteration 2 is',Q1)
print('The value of F(R_0) is',FR_1)
print('The z-value would be',z1)
print('The R_0 value would be',R1)

The optimal quantity for iteration 2 is 80.43426197663044
The value of F(R_0) is 0.9569102167982337
The z-value would be 1.7159042627408196
The R_0 value would be 115.13533670324544


The values of R_0 and R_1 match in iteration 1 and 2, hence further iterations will be terminated. Therefore the value of (Q,R) would be (80,115)

In [5]:
# Assume s = R
R = 115
Q = 80
S = R + Q
s = R

print('The value of s is',s)
print('The value of S is',S)

The value of s is 115
The value of S is 195


Therefore the appropriate value of (s,S) is (115,195)

(b) Suppose that the demands during the months of January to June were

So s = 196 and S = 115


Month|Demand|End of Month Inventory |Order|
-----|-----|-----|-----
December|-----|26|196-26=170
January|37|196-37=159|None
February|33|159-33=126|None
March|26|126-26=100|196-100=96
April|31|196-31=165|None
May|14|165-14=151|None
June|40|151-40=111|196-111=85

If the starting inventory in January was 26 cans of paint, determine the number of units
of paint ordered in each of the months January to June, following the (s; S) policy you
found above.

# Q2
Office-Co sells secretarial chairs. Annual demand for the chairs is normally distributed witha mean of 1040 chairs and a standard deviation of 51 chairs. Office-Co checks inventory at the end of each week (typically a Saturday evening) and if necessary, puts in an order to its supplier. It costs \\$100 to place an order and the lead time to receive an order is two weeks. Office-Co estimates that if a customer wants a chair when the company is out-of-stock, the loss-of-goodwill cost is \\$50 per chair. Each chair costs \\$60 and is sold for \\$100. The holding cost of the chairs is based on a 30% annual interest rate. (There are 52 weeks in a year)

(a) Discuss why it is appropriate to use an (s; S) inventory control policy in this case. Under what condition would you use a basestock (order-up-to) policy?

In [38]:
# Given

d    = 1040  # mean demand per year
d_sd = 51    # Std deviation per year
c    = 60    # each chair cost (unit cost)
K    = 100   # Order placement cost
r    = 100   # reselling cost
p    = 50    # stock-out cost (loss of goodwill)
i    = 0.3   # annual interest rate
x    = 1
lt   = 2     # Lead time

# Mean of demand during lead time
mu = round((d/52)*2, 2)

# Standard deviation of demand during lead time
sd = round(d_sd * np.sqrt((1/52)*2), 2)

print("The mean during lead time is", mu)
print("\nThe standard deviation during lead time is", sd)

The mean during lead time is 40.0

The standard deviation during lead time is 10.0


Office- co checks inventory at the end of each week and orders only if necessary.

Therefore (s,S) policy is appropriate as K>0 and the firm will order when the inventory position becomes sufficiently low.

(b) Using the (Q;R) solution, determine the appropriate values of the (s; S) inventory control policy. Use the EOQ value for the ordering quantity (i.e., no need for iterations).

In [39]:
# Optimal order quantity EOQ

EOQ = np.sqrt((2*K*d)/(i*c))

# To find z
FR1 = round(1-((EOQ*i*c)/(p*d*x)), 4) 
z1  = round(st.norm.ppf(FR1), 2) 

# Reorder point
R1 = sd*z1 + mu

# Inventory control policy
s = round(R1)
S = round(EOQ+R1)

print("EOQ value is ", EOQ,"units")
print("F(R1) and z for the order quantity is \nF(R1)=",FR1,"\nz=",z1)
print('R1= ', round(R1))
print("(s, S) is",(s,S))

EOQ value is  107.49676997731399 units
F(R1) and z for the order quantity is 
F(R1)= 0.9628 
z= 1.78
R1=  58
(s, S) is (58, 165)


The appropriate inventory control policy (s, S) is ( 58 , 165 )

(c) Suppose the company wants to have, on average, at most eight weeks with stockouts every year. Would the above calculated policy satisfy this requirement? Justify your answer
with calculations.

In [21]:
# The demand is not satisfied for 8 weeks
alpha = (52-8)/52
print("Type I service: alpha=",alpha,"or", alpha*100,"%")

Type I service: alpha= 0.8461538461538461 or 84.61538461538461 %


This Type I service obtained (alpha=0.8461) is less than that obtained in part b which is alpha(FR1)=0.9628, Hence the policy with 0.8461 is less stringent and would satisfy the requirement

(d) Can you propose a policy that satisfies the requirement of \at most eight weeks with stockouts per year" with less cost than the one calculated in part (b)? How much can you improve on the expected inventory holding cost?

In [34]:
# Total cost for part b with (s=58 and S=165) or (Q=107, R=58)

s    = 58
S    = 165
Q    = 107
R    = 58
n_R1 = 0.15
TC_Q = i*c*((0.5*Q)+R-mu) + (K*d*x)/Q + (p*d*n_R1)/Q

# find optimal policy with lower cost than part b for alpha=0.8461
FR1 = 0.8461 

#To find z 
z1 = round(st.norm.ppf(FR1), 2)
R1 = sd*z1 + mu 

# Optimal order point
Q1 = np.sqrt((2*d*x*(K+p*n_R1))/(i*c))

# To find z 
FR2 = round(1-((Q1*i*c)/(p*d*x)), 4) 
z2  = round(st.norm.ppf(FR2), 2) 

#Reorder point
R2 = sd*z2 + mu

# From
z2  = 1.76
Lz2 = 0.0158  # from standard normal table

# Expected shortage
nR2 = round((sd*Lz2),4)

# Optimal order point
Q2 = np.sqrt((2*d*x*(K+p*nR2))/(i*c))
TC_Q2 = i*c*((0.5*Q2)+R2-mu) + (K*d*x)/Q2 + (p*d*nR2)/Q2

# For policy in partb
T_HC1 = (i*c*(R-mu+(0.5*Q)))

# for the new policy
T_HC2 = (i*c*(R2-mu+(0.5*Q2)))

print("z1=",z1)
print("R1=",R1)
print('R2= ', R2)
print("Q1=",Q1)
print("Q2=",Q2)
print("n(R2)=",nR2)
print("Iteration 2") 
print("F(R2) and z2 for the order quantity is \nF(R2)=",FR2,"\nz=",z2)
print("The total cost for the policy in part b is",TC_Q)
print("The total cost for the new policy is",TC_Q2)
print("Total holding cost for policy in part b is",T_HC1,"\nTotal holding cost for new policy is", T_HC2) 
print("Difference in holding cost is",T_HC2-T_HC1)

z1= 1.02
R1= 50.2
R2=  57.7
Q1= 111.45502331533659
Q2= 111.66218896495108
n(R2)= 0.158
Iteration 2
F(R2) and z2 for the order quantity is 
F(R2)= 0.9614 
z= 1.76
The total cost for the policy in part b is 2331.8598130841124
The total cost for the new policy is 2328.51940136912
Total holding cost for policy in part b is 1287.0 
Total holding cost for new policy is 1323.5597006845596
Difference in holding cost is 36.55970068455963


The Total cost for the new policy (Q=111, R=57) in part d is less than that in part b (Q=107,R=58). But the holding cost can be improved by $35

(e) Office-co is considering investing in technology so that they can move from an (s; S) policy to a (Q;R) policy. Calculate a (Q;R) policy by performing a single iteration (Start with Q0=EOQ and perform one iteration to obtain the corresponding R1 and Q1).

In [33]:
# Optimal order quantity EOQ
EOQ=np.sqrt((2*K*d)/(i*c))

# To find z 
FR1=round(1-((EOQ*i*c)/(p*d*x)), 4) 
z1=round(st.norm.ppf(FR1), 2) 

# Reorder point
R1=sd*z1 + mu

# Iteration1
z1 = 1.78
Lz = 0.0152  # from standard normal table

# Expected shortage
nR1=round((sd*Lz),4)

# Optimal order point
Q1=np.sqrt((2*d*x*(K+p*nR1))/(i*c))

print("EOQ value is ", EOQ,"units")
print("F(R1) and z for the order quantity is \nF(R1)=",FR1,"\nz=",z1)
print('R1= ', round(R1))
print("n(R1)=",nR1)
print("Q1=",Q1)

EOQ value is  107.49676997731399 units
F(R1) and z for the order quantity is 
F(R1)= 0.9628 
z= 1.78
R1=  58
n(R1)= 0.152
Q1= 111.50685081096039


(Q, R) policy by performing a single iteration (111, 58)

(f) What is the Type I and Type II service levels obtained by the (Q;R) policy that you have calculated in Part (e)?

In [28]:
# Type I service calculated in part (c)
beta = 1-(nR1/Q1)
print("Type I service is alpha = 0.9626, i.e. 96.26 % ")
print("Type II service is: Beta = ",beta,"i.e.",beta*100,"%")

Type I service is alpha = 0.9626, i.e. 96.26 % 
Type II service is: Beta =  0.998636855055142 i.e. 99.8636855055142 %


(g) Discuss how you would justify the costs of this technology. What kinds of systems would enable the company to \move from an (s; S) policy to a (Q,R) policy" ?

In [29]:
# justify costs of this techology
print("The total cost for the policy in part b (s,S) is",TC_Q)
print("The total cost for the new policy in part d (Q,R) is",TC_Q2)

The total cost for the policy in part b (s,S) is 2331.8598130841124
The total cost for the new policy in part d (Q,R) is 2328.51940136912


Q,R policy gives a lower total cost, lower fixed cost but increased holding cost. So, if a lower total cost and lower fixed cost are more favourable to the company then it will move from (s,S) to (Q,R)

# Q3
Quick Print Inc. uses plain and three-hole punched paper for copying needs. Demand for each paper type is highly variable. Weekly demand for the plain paper is estimated to be normally distributed with mean 100 and standard deviation 65 (measured in boxes). Each week, a replenishment order is placed to the paper factory and the order arrives five weeks later. All copying orders that cannot be satisfed immediately due to lack of paper are back-ordered. The inventory holding cost is about \\$1 per box per year. Quick Print estimates that the backorder
penalty cost is \\$20 per box. Assume that there are 4.33 weeks per month.

(a) Suppose that Quick Print decides to establish an order-up-to level of 700 for plain paper. At the start of this week, there are 523 boxes in inventory and 180 boxes on order. How much will Quick Print order this week?

In [10]:
# Given 
IP   = 700   # Target inventory position
OH   = 523   # Boxes on-hand inventory
OO   = 180   # Boxes On order 

CI = OH + OO 
print('\n', CI)
print('\n The current inventory position is',CI,'which is more than the target inventory position of 700, therefore there will be no order this week\n')


 703

 The current inventory position is 703 which is more than the target inventory position of 700, therefore there will be no order this week



(b) Again, assume that Quick Print uses an order-up-to level of 700.

i. What is Quick Print's Type I service level?

ii. What is Quick Print's expected on-hand inventory?

In [11]:
BSL  = 700    # base stock level (order upto level)
r    = 1      # in weeks
L    = 5      # in weeks
u3   = 100
std3 = 65

z = (700 - ((r+L)*u3))/(std3*np.sqrt(r+L))
print('The z value would be',z)

#To find average inventory level(on-hand inventory)
I=(r*d*0.5)+((z*std3)*(np.sqrt(r+L))) 
print("Expected on-hand inventory is",I)

The z value would be 0.6280742930213278
Expected on-hand inventory is 620.0


Therefore after observing the standard normal table, we determine the type I service level to be 73.24 %

(c) What is Quick Print's optimal order-up-to level for plain paper if Quick Print operates with a 85% in-stock probability (i.e., Type I service level)?

In [12]:
z    = 1.04 # z-value corresponding to 85% stock probability (Type I service level)
u3   = 100  # average weekly demand
std3 = 65   # standard deviation of weekly demand
P    = 1    # Replenishment time in weeks
L    = 5    # Lead time in weeks

Ss = z*std3*(np.sqrt(P+L))
Ot = u3*(P+L) + Ss

print('\nThe safety stock is',Ss)
print('\nQuick Prints optimal target inventory position will be ',Ot)


The safety stock is 165.58550661214284

Quick Prints optimal target inventory position will be  765.5855066121428
